File: splitiq/src/splitiq/ratio.py
"""Optimal train/test split ratio (Joseph 2022)."""

from __future__ import annotations

import numpy as np

from splitiq._convert import DataLike, to_julia_data
from splitiq._julia import _translate_error, julia


def optimal_split_ratio(
    x: DataLike, y: DataLike, *, method: str = 'simple', degree: int = 2
) -> float:
    """Optimal fraction of data to assign to the test set (Joseph 2022, Eq. 11).

    Args:
        x: The predictors: a numpy array-like (1-D or 2-D) or a pandas
            DataFrame.
        y: The response; must have the same number of observations as `x`.
        method: ``'simple'`` counts the encoded predictor columns of `x`
            (after preprocessing: categorical columns Helmert-encoded,
            constant columns dropped) plus one for the intercept.
            ``'regression'`` is the paper's model-selection strategy for an
            unknown model; not implemented in this release.
        degree: Polynomial expansion degree for ``method='regression'``.

    Returns:
        The optimal test-set fraction, ``1 / (sqrt(p) + 1)``.

    Raises:
        ValueError: If `method` is not ``'simple'`` or ``'regression'``, or
            if `x` and `y` have a different number of observations.
        NotImplementedError: If `method` is ``'regression'`` (not yet
            implemented).
    """
    from juliacall import JuliaError

    jl = julia()
    x_julia = to_julia_data(x)
    y_julia = np.asarray(y, dtype=np.float64)
    method_symbol = jl.Symbol(method)
    try:
        with _translate_error():
            value = jl.optimal_split_ratio(x_julia, y_julia, method=method_symbol, degree=degree)
    except JuliaError as exc:
        if jl.isa(exc.exception, jl.ErrorException):
            message = str(jl.sprint(jl.showerror, exc.exception))
            raise NotImplementedError(message) from exc
        raise
    return float(value)

File: src/ratio.jl
"""
Optimal splitting ratio from Joseph (2022), "Optimal Ratio for Data Splitting"
(Statistical Analysis and Data Mining 15(4), 531-538). The optimal testing
fraction is γ = 1/(√p + 1) (Eq. 11), where p is the number of parameters of
the linear regression model f(x)'β expected to explain the data well,
including the intercept (Section 2.2).
"""

"""
    optimal_split_ratio(x, y; method = :simple, degree = 2) -> Float64

Optimal fraction of data to assign to the test set, following the
γ* = 1/(√p + 1) result of Joseph (2022, Eq. 11).

- `method = :simple`: p is taken as the number of encoded predictor columns
  of `x` (after [`preprocess`](@ref): categorical columns Helmert-encoded,
  constant columns dropped) plus one, for the intercept.
- `method = :regression`: the paper's practical strategy (Section 3) for
  when the model is unknown — expand `x` into a larger feature set (e.g.
  polynomial terms up to `degree`), fit a linear regression on the full
  data, and select p via a model-selection criterion such as Cp or AIC.
  Not implemented in this release; calling with this method raises an
  error rather than silently falling back to `:simple`.

`x` may be an `AbstractVector`, `AbstractMatrix`, or `DataFrame`; `y` must
have the same number of observations as `x`.
"""
function optimal_split_ratio(x, y; method::Symbol = :simple, degree::Int = 2)
  method in (:simple, :regression) ||
    throw(ArgumentError("method must be :simple or :regression, got :$method"))
  nobs = x isa AbstractVector ? length(x) : size(x, 1)
  length(y) == nobs ||
    throw(ArgumentError("x and y must have the same number of observations"))

  if method === :regression
    error(
      "method = :regression is not implemented yet; the model-selection " *
      "based estimation procedure of p from Joseph (2022, Section 3) lands " *
      "in a later release",
    )
  end

  p = size(preprocess(x), 2) + 1
  return 1 / (sqrt(p) + 1)
end
